scheduler/reconciler/reconcile_node.go

```diff
@@ -378,6 +437,46 @@ func (nr *NodeReconciler) diffSystemAllocsForNode(
 	return result, deploymentComplete
 }
 
+func (nr *NodeReconciler) cancelUnneededCanaries(allocs []*structs.Allocation, required map[string]*structs.TaskGroup) []AllocTuple {
+	stop := []AllocTuple{}
+
+	// Cancel non-promoted canaries from previous deployments
+	if nr.DeploymentOld != nil {
+		for oldDeployTGName, dstate := range nr.DeploymentOld.TaskGroups {
+			if !dstate.Promoted {
+				for _, a := range allocs {
+					if slices.Contains(dstate.PlacedCanaries, a.Name) {
+						stop = append(stop, AllocTuple{
+							Name:      a.Name,
+							TaskGroup: required[oldDeployTGName],
```

Comment:
Are we guaranteed the old task group name exists in `required`? What happens if we drop a task group in a new version of a job?